test_conformance/extensions/cl_khr_command_buffer/command_buffer_queue_substitution.cpp

```diff
@@ -0,0 +1,321 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "basic_command_buffer.h"
+#include "procs.h"
+
+#include <vector>
+
+namespace {
+
+////////////////////////////////////////////////////////////////////////////////
+// Command-queue substitution tests which handles below cases:
+// -substitution on queue without properties
+// -substitution on queue with properties
+// -simultaneous use queue substitution
+
+template <bool prop_use, bool simul_use>
+struct SubstituteQueueTest : public BasicCommandBufferTest
+{
+    SubstituteQueueTest(cl_device_id device, cl_context context,
+                        cl_command_queue queue)
+        : BasicCommandBufferTest(device, context, queue),
+          properties_use_requested(prop_use), user_event(nullptr)
+    {
+        simultaneous_use_requested = simul_use;
+        if (simul_use) buffer_size_multiplier = 2;
+    }
+
+    //--------------------------------------------------------------------------
+    bool Skip() override
+    {
+        if (properties_use_requested && queue == nullptr) return true;
+
+        return (simultaneous_use_requested && !simultaneous_use_support)
+            || BasicCommandBufferTest::Skip();
+    }
+
+    //--------------------------------------------------------------------------
+    cl_command_queue CreateCommandQueueWithProperties(cl_int& error)
+    {
+        cl_command_queue ret_queue = nullptr;
+        cl_queue_properties_khr device_props = 0;
+
+        error = clGetDeviceInfo(device, CL_DEVICE_QUEUE_PROPERTIES,
+                                sizeof(device_props), &device_props, nullptr);
+        test_error_ret(error,
+                       "clGetDeviceInfo for CL_DEVICE_QUEUE_PROPERTIES failed",
+                       nullptr);
+
+        using PropPair = std::pair<cl_queue_properties_khr, std::string>;
+
+        auto check_property = [&](const PropPair& prop) {
+            if (device_props & prop.first)
+            {
+                log_info("Queue property %s supported. Testing ... \n",
+                         prop.second.c_str());
+                ret_queue =
+                    clCreateCommandQueue(context, device, prop.first, &error);
+            }
+            else
+                log_info("Queue property %s not supported \n",
+                         prop.second.c_str());
+        };
+
+        // in case of extending property list in future
+        std::vector<PropPair> props = {
+            ADD_PROP(CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE),
+            ADD_PROP(CL_QUEUE_PROFILING_ENABLE)
+        };
+
+        for (auto&& prop : props)
+        {
+            check_property(prop);
+            test_error_ret(error, "clCreateCommandQueue failed", ret_queue);
+            if (ret_queue != nullptr) return ret_queue;
```

Comment:
according to issue description:

>Create a command-buffer recorded to a single queue with additional properties (e.g profiling enabled) set

so this code will try to create either out-of-order or profiling queue depending on device capabilities

Reply:
If we're only going to test one of the queue properties (which is fine) then I have a slight preference for (A), but only a slight preference.  The nice thing about (A) is that if somebody modifies this test in the future then it can be tested on any implementation, whereas (B) would require testing on implementations with and without out-of-order queue support.

This does assume that we will have tests for out-of-order queues elsewhere so we won't need them here but I think that's a safe assumption.

Reply:
Cool, let's go with A to keep things simple since we do indeed have out-of-order queue testing elsewhere

Reply:
Done